custom-types-quick-guide.md

```diff
@@ -0,0 +1,245 @@
+      since: 2.7
+    - name: port
+      type: int
+      nullable: false
+      since: 2.7
+- name: ServiceBConfig
+  since: 2.7
+  params:
+    - name: address
+      type: String
+      nullable: false
+      since: 2.7
+    - name: addressParser
+      type: Data
+      nullable: false
+      since: 2.7
+```
+
+Open `schema/protocol-schema.json` and append to the array `definitions.param.properties.type.enum`
+the string literals: `ServiceAConfig` and `ServiceBConfig`.
+
+Now we need to define the corresponding domain types in Java that these custom types will map-to.
+Let us consider the following existing `ServiceAConfig` and `ServiceBConfig` classes that already
+exist and which cannot be modified for reasons of backwards compatibility.
+
+```java
+public final class ServiceAConfig {
+    private final String address;
+    private final int port;
+
+    public ServiceAConfig(String address, int port) {
+        this.address = address;
+        this.port = port;
+    }
+
+    public String getAddress() {
+        return address;
+    }
+
+    public int getPort() {
+        return port;
+    }
+}
+
+public final class ServiceBConfig {
+    private final String address;
+    private final AddressParser parser;
+
+    public ServiceBConfig(String address, AddressParser parser) {
+        this.address = address;
+        this.parser = parser;
+    }
+
+    public String getAddress() {
+        return address;
+    }
+
+    public AddressParser getParser() {
+        return parser;
+    }
+}
+```
+
+`ServiceAConfig` has a simple mapping to its Java domain type; however, `ServiceBConfig` uses `Data`
+as its type for `parser` and not `AddressParser`: this is to show an example where a configuration
+carries something that is specific to the target language -- in this case an `AddressParser` which
+cannot be represented in the client protocol, hence it being `Data`. You don't need to know anything
+about `AddressParser` but that it implements `IdentifiableDataSerializable` and can therefore be
+deserialized. For `ServiceBConfig` we must create an intermediate domain type, a so-called holder.
+
+```java
+public final class ServiceBConfigHolder {
+    private final String address;
+    private final Data addressParser;
+
+    public ServiceBConfigHolder(String address, Data addressParser) {
+        this.address = address;
+        this.addressParser = addressParser;
+
+
+    public String getAddress() {
+        return address;
+    }
+
+    public Data getAddressParser() {
+        return addressParser;
+    }
+}
+```
+
+The `ServiceBConfigHolder` is the same as `ServiceBConfig` Java class with the exception that we
+have made `addressParser` now the opaque `Data` type.
+
+With the boilerplate in-place we can now add the following to `java/__init__.py`'s
+`_java_types_common` dictionary:
+
+```python
+    "ServiceAConfig": "com.hazelcast.config.ServiceAConfig",
+    "ServiceBConfig": "com.hazelcast.config.ServiceBConfigHolder"
```

Comment:
this is something we already discussed in your PR, we skipped there not the create friction but for the readme I am note sure. 

so what about naming this ServiceBConfigHolder instead because this is what we did for others in the past.? 

It may make the README a bit confusing but I think we can fix that by describing why we need a holder before using a holder, basically one of the first things to describe.

Reply:
yeah I was talking about describing it earlier if we use holder name because protocol yaml will have it. 

Thing is, the holder is not something user sees anyway, my rename suggestion is just to follow our convention so far.